src/main/scala/vpio/IO.scala

```diff
@@ -60,25 +128,51 @@ object IO extends CommandApp(
       val layout: LayoutDefinition =
         ZoomedLayoutScheme.layoutForZoom(14, WebMercator.worldExtent, 512)
 
-      osm.fromORC(orc) match {
+      val df = ss.read.orc(orc)
+      //      val df = ss.read.orc("s3://osm-pds/planet/planet-latest.orc")
+
+      // val southAmericaBeforeJune =
+      //   df.where("timestamp <= '2017-06-01 00:00:00.0' AND (type != 'node' OR (lon <= -34.767608642578125 AND lon >= -81.33865356445312 AND lat <= 12.623252653219012 AND lat >= -55.98609153380838))")
+
+      // val southAmerica =
+      //   df.where("type != 'node' OR (lon <= -34.767608642578125 AND lon >= -81.33865356445312 AND lat <= 12.623252653219012 AND lat >= -55.98609153380838)")
+
+      val targetDf = df//.repartition(1000)
+
+      // Test log clip fail
+      val ff = Feature(Point(1,1), vectorpipe.util.Tree(osm.ElementData(osm.ElementMeta(1L, "Asdf", "asdf", 2L, 3L, "asdf", true), Map(), None), Seq(vectorpipe.util.Tree(osm.ElementData(osm.ElementMeta(2L, "Asdf", "asdf", 2L, 3L, "asdf", true), Map(), None), Seq()))))
+
+      Util.logClipFail(Extent(0, 0, 1, 1), ff)
+
+
+      osm.fromDataFrame(targetDf) match {
         case Left(e) => println(e)
         case Right((ns,ws,rs)) => {
+          val numPartitions = 10000
+          val nodePartitioner = new HashPartitioner(numPartitions)
+          val wayPartitioner = new HashPartitioner(numPartitions)
+
+          /* Reproject nodes */
+          val reprojectedNodes =
+            ns.partitionBy(nodePartitioner).mapPartitions({ partition =>
+              val transform = Transform(LatLng, WebMercator)
+              partition.map { case (nodeId, node) =>
+                val (lon, lat) = transform(node.lon, node.lat)
+                (nodeId, node.copy(lon = lon, lat = lat))
+              }
+            }, preservesPartitioning = true)
 
           /* Assumes that OSM ORC is in LatLng */
-          val latlngFeats: RDD[osm.OSMFeature] =
-            osm.toFeatures(ns.repartition(100), ws.repartition(10), rs)
-
-          /* Reproject into WebMercator, the default for VTs */
-          val wmFeats: RDD[osm.OSMFeature] =
-            latlngFeats.map(_.reproject(LatLng, WebMercator))
+          val feats: RDD[osm.OSMFeature] =
+            osm.toFeatures(reprojectedNodes, ws.partitionBy(wayPartitioner), rs)
 
           /* Associated each Feature with a SpatialKey */
           val fgrid: RDD[(SpatialKey, Iterable[osm.OSMFeature])] =
-            VectorPipe.toGrid(Clip.byHybrid, VectorPipe.log4j, layout, wmFeats)
+            VectorPipe.toGrid(Clip.byHybrid, Util.logClipFail, layout, feats, new HashPartitioner(numPartitions))
 
           /* Create the VectorTiles */
           val tiles: RDD[(SpatialKey, VectorTile)] =
-            VectorPipe.toVectorTile(Collate.withoutMetadata, layout, fgrid)
+            VectorPipe.toVectorTile(Collate.byAnalytics, layout, fgrid).cache()
```

Comment:
byAnalytics on South America choked for me.

Reply:
:(